python/tests/kat/t_tcpmapping.py

```diff
@@ -264,3 +265,735 @@ def check(self):
+                    sni=True, headers={"Host": "tls-context-host-2"}, ca_cert=TLSCerts["tls-context-host-2"].pubcert)
+
+    def check(self):
+        assert self.results[0].json["backend"] == self.target.path.k8s
+        assert self.results[0].json["request"]["tls"]["enabled"] == False
+
+
+class TCPMappingAddressTest(AmbassadorTest):
+    extra_ports = [ 6789, 6790 ]
+    target: ServiceType
+
+    def init(self) -> None:
+        self.target = HTTP()
+
+    def manifests(self) -> str:
+        return format('''
+---
+apiVersion: getambassador.io/v2
+kind: TCPMapping
+metadata:
+  name: {self.name.k8s}-local-only
+spec:
+  ambassador_id: [ {self.ambassador_id} ]
+  port: 6789
+  address: 127.0.0.1
+  service: {self.target.path.fqdn}:80
+---
+apiVersion: getambassador.io/v2
+kind: TCPMapping
+metadata:
+  name: {self.name.k8s}-proxy
+spec:
+  ambassador_id: [ {self.ambassador_id} ]
+  port: 6790
+  service: localhost:6789
+''') + super().manifests()
+
+    def queries(self):
+        # Check that it only bound to localhost and doesn't allow external connections.
+        yield Query(self.url("", port=6789),
+                    error=[ 'connection reset by peer', 'EOF', 'connection refused' ])
+        # Use a second mapping that proxies to the first to check that it was even created.
+        yield Query(self.url("", port=6790))
+
+    def check(self):
+        assert self.results[1].json["backend"] == self.target.path.k8s
+        assert self.results[1].json["request"]["tls"]["enabled"] == False
+
+
+class TCPMappingWeightTest(AmbassadorTest):
+    extra_ports = [ 6789 ]
+    target70: ServiceType
+    target30: ServiceType
+
+    def init(self) -> None:
+        self.target70 = HTTP(name="tgt70")
+        self.target30 = HTTP(name="tgt30")
+
+    def manifests(self) -> str:
+        return format('''
+---
+apiVersion: getambassador.io/v2
+kind: TCPMapping
+metadata:
+  name: {self.name.k8s}-70
+spec:
+  ambassador_id: [ {self.ambassador_id} ]
+  port: 6789
+  service: {self.target70.path.fqdn}:80
+  weight: 70
+---
+apiVersion: getambassador.io/v2
+kind: TCPMapping
+metadata:
+  name: {self.name.k8s}-30
+spec:
+  ambassador_id: [ {self.ambassador_id} ]
+  port: 6789
+  service: {self.target30.path.fqdn}:80
+  weight: 30
+''') + super().manifests()
+
+    def queries(self):
+        for i in range(1000):
+            yield Query(self.url("", port=6789))
+
+    def check(self):
+        counts: Dict[str, int] = {}
+        for result in self.results:
+            backend = result.json["backend"]
+            counts[backend] = counts.get(backend, 0)+1
+        assert counts[self.target70.path.k8s]+counts[self.target30.path.k8s] == 1000
+        # Probabalistic, margin might need tuned
+        margin = 150
+        assert abs(counts[self.target70.path.k8s]-700) < margin
+        assert abs(counts[self.target30.path.k8s]-300) < margin
+
+
+# TODO: Add tests for all of the config knobs for the upstream connection:
```

Comment:
Rather than have a TODO at the bottom of a file, can you please create an issue for this and reference this PR?  

As we have seen from t_stats being disabled and the numerous other spots in the code base  where these get completely lost and ignored for years. An issue has the same potential but has the benefit of being more visible to the community, the support team, customer experience team, leadership team and ourselves.

Reply:
+1

Reply:
I've filed https://github.com/emissary-ingress/emissary/issues/4516

That said, it's a problem on its own that TODOs don't get attention.